src/main/java/io/aiven/kafka/auth/LocklessAivenAclAuthorizer.java

```diff
@@ -126,8 +126,8 @@ private boolean reloadConfigIfNecessary() {
                 LOGGER.info("Reloading ACL configuration {}", configFile);
                 try {
                     final List<AivenAcl> newAclEntries = jsonReader.read();
-                    verdictCache.clear();
                     aclEntries.set(newAclEntries);
+                    verdictCache.clear();
```

Comment:
does this really fix the race condition? Isn't reordering allowed in the java memory model?

Reply:
There is an issue that I didn't thought before. As the comment above explains, I thought it was okay to have this ordering:

| T1 | T2 | details
set | | T1 updates the cache with the new rules
| checkAcl | T2 reads a stale verdict, assumed to be okays as per the comment
clear | | the cache is cleared, from now on the verdict will be in sync with the rules 

But this is also possible:

| T1 | T2 | details
| clear | | cache clearing is reordered (its side effect appear first to T2)
| | checkAcl | T2 inserts a new entry in the cache based on the *old* rules
set | | T1 updates the rules

The side effect is that the cache may have an invalid entry until the rules are reloaded

Reply:
It is not possible because `AtomicReference#set` has the memory effects of writing (assigning) a volatile variable.